File: examples/multilinguagem/python/aws.py
"""
Como este lab constroi clients boto3.

Regra que atravessa as tres linguagens: NENHUMA credencial no construtor.
O que muda entre local, EKS, ECS e Lambda e o AMBIENTE, nunca o codigo.
"""

import os

import boto3
from botocore.config import Config

# Retry adaptativo: o modo "standard"/"adaptive" do botocore ja trata
# throttling e erros transitorios. Sem isto, um pico no STS ou na SQS vira
# excecao no seu loop em vez de uma nova tentativa silenciosa.
_CONFIG = Config(retries={"max_attempts": 5, "mode": "standard"})


def endpoint() -> str | None:
    """
    O endpoint do emulador, quando existe.

    boto3 recente ja le AWS_ENDPOINT_URL sozinho, mas a versao em que isso
    passou a valer varia entre SDKs (o .NET so ganhou equivalente bem depois).
    Ler a variavel na mao e passar explicitamente deixa as tres linguagens
    deste lab com o MESMO comportamento, independente da versao instalada.

    Em producao a variavel simplesmente nao existe e isto devolve None -
    o SDK resolve o endpoint real da regiao.
    """
    return os.environ.get("AWS_ENDPOINT_URL", "").strip() or None


def cliente(servico: str):
    """
    Um client boto3 para o servico pedido.

    Repare no que NAO tem aqui: aws_access_key_id, aws_secret_access_key,
    profile_name, nenhum if de ambiente. O botocore percorre a propria cadeia
    de provedores (guia 03, secao 2) e para no primeiro que responder.

    E repare que quem chama guarda o resultado: client boto3 e caro de criar
    e o cache de credencial mora dentro dele. Criar um por mensagem forca um
    AssumeRoleWithWebIdentity novo a cada vez - latencia, throttling do STS e
    um CloudTrail impossivel de ler.
    """
    return boto3.client(servico, endpoint_url=endpoint(), config=_CONFIG)

File: examples/multilinguagem/dotnet/Aws.cs
// Como este lab constroi clients do AWS SDK for .NET.
//
// Regra que atravessa as tres linguagens: NENHUMA credencial no construtor.
// O que muda entre local, EKS, ECS e Lambda e o AMBIENTE, nunca o codigo -
// e e por isso que este mesmo binario roda nos quatro lugares sem um if.

using Amazon.Runtime;
using Amazon.SecretsManager;
using Amazon.SecurityToken;
using Amazon.SimpleNotificationService;
using Amazon.SQS;

namespace Lab;

public static class Aws
{
    /// <summary>
    /// O endpoint do emulador, ou null em AWS de verdade.
    ///
    /// Diferente do boto3 e do SDK Go, o AWS SDK for .NET so ganhou suporte a
    /// AWS_ENDPOINT_URL bem depois. Ler a variavel na mao e atribuir a
    /// ServiceURL deixa as tres linguagens deste lab com o mesmo
    /// comportamento, independente da versao do pacote.
    /// </summary>
    public static string? Endpoint =>
        Environment.GetEnvironmentVariable("AWS_ENDPOINT_URL") is { Length: > 0 } e
            ? e.Trim()
            : null;

    private static string Regiao =>
        Environment.GetEnvironmentVariable("AWS_REGION")
        ?? Environment.GetEnvironmentVariable("AWS_DEFAULT_REGION")
        ?? "us-east-1";

    /// <summary>
    /// Aplica o endpoint local, quando ele existe.
    ///
    /// AuthenticationRegion e obrigatorio junto com ServiceURL: sem regiao o
    /// SDK nao consegue montar a assinatura SigV4. E o mesmo sintoma do guia
    /// 03 §5 - "No RegionEndpoint or ServiceURL configured" nao e problema de
    /// IAM, e regiao faltando. Sao duas cadeias independentes.
    /// </summary>
    private static T Ajustar<T>(T config) where T : ClientConfig
    {
        if (Endpoint is { } url)
        {
            config.ServiceURL = url;
            config.AuthenticationRegion = Regiao;
        }
        return config;
    }

    // Um construtor por servico. Nenhum recebe AWSCredentials.
    //
    // Guarde o client que sai daqui - o cache de credenciais mora DENTRO do
    // objeto de credenciais, que mora dentro do client. Criar um client por
    // mensagem joga esse cache fora e forca um AssumeRoleWithWebIdentity novo
    // a cada vez: latencia, throttling do STS e milhares de eventos no
    // CloudTrail por hora (guia 03 §4).
    public static IAmazonSQS Sqs() => new AmazonSQSClient(Ajustar(new AmazonSQSConfig()));

    public static IAmazonSimpleNotificationService Sns() =>
        new AmazonSimpleNotificationServiceClient(Ajustar(new AmazonSimpleNotificationServiceConfig()));

    public static IAmazonSecurityTokenService Sts() =>
        new AmazonSecurityTokenServiceClient(Ajustar(new AmazonSecurityTokenServiceConfig()));

    public static IAmazonSecretsManager SecretsManager() =>
        new AmazonSecretsManagerClient(Ajustar(new AmazonSecretsManagerConfig()));
}
